app/routes/resources+/images.tsx

```diff
@@ -0,0 +1,91 @@
+import { invariantResponse } from '@epic-web/invariant'
+import { getImgResponse } from 'openimg/node'
+import { prisma } from '#app/utils/db.server.ts'
+import { getDomainUrl } from '#app/utils/misc.tsx'
+import { getSignedGetRequestInfo } from '#app/utils/storage.server.ts'
+import { type Route } from './+types/images.ts'
+
+export async function loader({ request }: Route.LoaderArgs) {
+	const domain = getDomainUrl(request)
+	const headers = new Headers()
+	headers.set('Cache-Control', 'public, max-age=31536000, immutable')
+	return getImgResponse(request, {
+		headers,
+		allowlistedOrigins: [domain, process.env.AWS_ENDPOINT_URL_S3].filter(
+			Boolean,
+		),
+		cacheFolder:
+			process.env.NODE_ENV === 'production'
+				? '/data/images'
+				: './tests/fixtures/openimg',
+		getImgSource: ({ params }) => {
+			if (params.src.startsWith('/resources')) {
+				const searchParams = new URLSearchParams(params.src.split('?')[1])
+				const userImageId = searchParams.get('userImageId')
+				if (userImageId) {
+					return handleUserImage(userImageId)
+				}
+				const noteImageId = searchParams.get('noteImageId')
+				if (noteImageId) {
+					return handleNoteImage(noteImageId)
+				}
+
+				// Fetch image from resource endpoint
+				return {
+					type: 'fetch',
+					url: domain + params.src,
+				}
+			}
+			if (URL.canParse(params.src)) {
+				// Fetch image from external URL; will be matched against allowlist
+				return {
+					type: 'fetch',
+					url: params.src,
+				}
+			}
+			// Retrieve image from filesystem (public folder)
+			if (params.src.startsWith('/assets')) {
+				// Files managed by Vite
+				return {
+					type: 'fs',
+					path: '.' + params.src,
+				}
+			}
+			// Fallback to files in public folder
+			return {
+				type: 'fs',
+				path: './public' + params.src,
+			}
+		},
+	})
+}
+
+async function handleUserImage(userImageId: string) {
+	const userImage = await prisma.userImage.findUnique({
+		where: { id: userImageId },
+		select: { objectKey: true },
+	})
+	invariantResponse(userImage, 'User image not found', { status: 404 })
+
+	const { url, headers } = getSignedGetRequestInfo(userImage.objectKey)
+	return {
+		type: 'fetch',
+		url,
+		headers,
+	}
```

Comment:
I need the fetch request you make to include these headers.

I also need `getImgSource` to support returning a promise.

Reply:
Available via `openimg` v0.2.0! 🎉